contracts/nexus-gateway/src/contract/execute.rs

```diff
@@ -1,10 +1,70 @@
-use cosmwasm_std::{to_json_binary, Response, Storage, WasmMsg};
+use axelar_wasm_std::msg_id::HexTxHashAndEventIndex;
+use axelar_wasm_std::nonempty;
+use cosmwasm_std::{BankMsg, HexBinary, MessageInfo, QuerierWrapper, Response, Storage, Uint256};
+use error_stack::{bail, ResultExt};
+use router_api::{Address, ChainName};
+use sha3::{Digest, Keccak256};
 
 use crate::error::ContractError;
+use crate::state::load_config;
 use crate::{nexus, state};
 
 type Result<T> = error_stack::Result<T, ContractError>;
 
+pub fn call_contract_with_token(
+    storage: &mut dyn Storage,
+    querier: QuerierWrapper,
+    block_height: u64,
+    info: MessageInfo,
+    destination_chain: ChainName,
+    destination_address: Address,
+    payload: HexBinary,
+) -> Result<Response<nexus::Message>> {
+    let config = load_config(storage)?;
+    let axelarnet_gateway: axelarnet_gateway::Client =
+        client::Client::new(querier, config.axelar_gateway).into();
+    let source_address: Address = info
+        .sender
+        .into_string()
+        .parse()
+        .expect("invalid sender address");
+    let token = match info.funds.as_slice() {
+        [token] => token.clone(),
+        _ => bail!(ContractError::InvalidToken(info.funds)),
+    };
+    // TODO: Retrieve the actual tx hash and event index from core, since cosmwasm doesn't provide it. Use the block height as the placeholder in the meantime.
+    let tx_hash = Uint256::from(block_height).to_be_bytes();
+    let event_index = 0u32;
+    let id = HexTxHashAndEventIndex::new(tx_hash, event_index);
```

Comment:
We can't use only block height, txs in the same block will have the same message id

Reply:
This is known, and this is the case in axelarnet-gateway for the moment as well. Another JIRA ticket is out there for implementing this in core so that contracts can retrieve.

Reply:
which is equally as bad cuz the counter isn't global and even worse hidden. Core should handle be able to return both the tx id and index so that the index can be unique in the context of the same tx. I do not want to extend the mistake.